### ISSUE TITLE
Polling Amazon SQS results in large quantities of memory being leaked

### ISSUE THREAD

**Issue by YorickPeterse** [rubinius/rubinius#3205]
While running an application that uses SQS (Amazon's queueing service) I noticed quite the memory increase over time:

![shepherd_rss_time](https://cloud.githubusercontent.com/assets/86065/5016808/958cd284-6aa9-11e4-872d-16bd3246ebdd.png)

This graph shows the application's total memory usage over time. Just before 2014-11-11 we deployed Rubinius 2.3 (the area with the 3 vertical bars). Directly after that memory usage almost doubled, although an increase in general was expected. However, the next peak was anything but expected. While more memory usage on Rubinius isn't uncommon a peak of almost 1,2GB indicated a serious problem. It was fairly clear to me there was a leak because:
1. This app has been running for months on MRI without leaking memory (ignoring the crap garbage collector and its heap bloat)
2. The application's own source code tries very hard _not_ to share any state, cache data globally, etc. In other words, either it's a Gem or Rubinius problem.

To figure out what was going on I employed a mix of my dtrace hacks (https://github.com/rubinius/rubinius/tree/dtrace-boxing) to trace allocations as well as showing verbose GC and JIT output (`-Xgc.show -Xgc.immix.debug -Xjit.show` to be exact). I also generated a bunch of Rbx heap dumps as well as running Rubinius under Massif (heap profiler part of Valgrind).

All these tools showed an interesting pattern, especially my dtrace hacks were tremendously helpful (not even kidding). Every 10 seconds or so (I didn't time it exactly) a whole bunch of allocations would occur, but without any GC runs following it. The interval matched the application's behaviour: every 10-or-so seconds it would poll Amazon SQS for a new job.

The interesting part here is the lack of GC runs. Assuming the AWS SDK doesn't keep data around (which I'm not sure about just yet) _somehow_ data would be allocated but never freed. Messing around with `ObjectSpace.find_object` and `ObjectSpace.find_references` didn't really help sadly.

However, based on what the application does (it's actually quite a simple application) I set up a standalone script that mimics part of its behaviour. Lo and behold, it also leaks memory:

``` ruby
gem 'aws-sdk', '~> 1.0'

require 'aws-sdk'
require 'thread'

AWS.config(:sqs_region => ENV['AWS_REGION'] || 'eu-west-1')

# AWS SDK lazy loads half the world, which isn't very thread-safe. Lets eager
# load what we need instead.
AWS.eager_autoload!(AWS::SQS)
AWS.eager_autoload!(AWS::Core)

Thread.abort_on_exception = true

stdout_lock = Mutex.new
num_threads = ENV['THREADS'] ? ENV['THREADS'].to_i : 10
queue_name  = ENV['QUEUE'] || 'example'

threads = num_threads.times.map do |n|
  Thread.new do
    stdout_lock.synchronize { puts "Starting thread #{n + 1}" }

    sqs   = AWS::SQS.new
    queue = sqs.queues.named(queue_name)

    queue.poll { }
  end
end

threads << Thread.new do
  loop do
    sleep 10

    memory = `ps -o rss= #{Process.pid}`.strip.to_f / 1024

    stdout_lock.synchronize { puts "Memory usage: #{memory.round(2)} MB" }
  end
end

%w{TERM INT QUIT}.each do |signal|
  trap(signal) do
    threads.each(&:kill)
  end
end

threads.each(&:join)
```

For this script you'll need to have an AWS account, a single SQS queue (default name is "example") and the following environment variables set:
- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`
- `AWS_REGION`

When running the script it should be fairly easy to notice it's leaking quite a bit of memory, especially when you crank up the amount of threads.

One interesting thing to note is that I've run my application using `-Xgc.immix.concurrent=false` which results in _way_ more stable memory usage:

![shepherd_rss_time_non_concurrent](https://cloud.githubusercontent.com/assets/86065/5017055/41878c86-6aab-11e4-9bf3-3081a7c70af7.png)

The first vertical line marks the deployment containing the `-Xgc.immix.concurrent=false` option. The second vertical line marks the deployment where I replaced this option with `-Xgc.young_max_bytes=8388608` as suggested per @brixen. This didn't really help much, so the third line marks the deployment that reverted back to using the Immix option.

**tl;dr:** Either Rubinius is not collecting objects properly, or the AWS SDK (or something it uses) is leaking memory without Rubinius being able to do something about it. Since I _suspect_ the former I'm reporting it here, but that's purely based on this problem never having occurred in the past.

**Comment by pickhardt** [comment]
@YorickPeterse based on your running of the script with both MRI and Rubinius, do you believe the memory leak is due to the AWS SQS library? If so, did you eventually figure out what to do to fix it? For instance, would upgrading to version 2 of the AWS-SDK gem fix it?

**Comment by YorickPeterse** [comment]
@pickhardt I can't say for certain if it's the AWS SDK itself, only that it manifests when using it. I've been unable to try version 2 yet as last time I checked it lacked a more refined SQS interface (that is, one closer to the interface of v1). I've not yet been able to find a fix otherwise.

**Comment by brixen** [comment]
In order to make performance work as efficient and effective as possible, the focus for Rubinius in the near term is on the following capabilities:

1. Instruction set
2. Debugger
3. Profiler
4. Just-in-time compiler
5. Concurrency
6. Garbage collector

Once these core capabilities are properly functioning, it will be possible to create a how-to guide to examining and fixing performance issues.
